packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
+        if (leftArrow)
+          newOffset = _baseOffset < _extentOffset ? _baseOffset : _extentOffset;
+        else if (rightArrow)
+          newOffset = _baseOffset > _extentOffset ? _baseOffset : _extentOffset;
+      }
+      onSelectionChanged(new TextSelection.fromPosition(
+        new TextPosition(offset: newOffset)),
+        this, SelectionChangedCause.keyboard);
+    }
+    return newOffset;
+  }
+
+  // Handles moving the cursor vertically as well as taking care of the
+  // case where the user moves the cursor to the end or beginning of the text
+  // and then back up or down.
+  int _handleVerticalArrows(bool upArrow, bool downArrow, bool shift, int newOffset) {
+    if (downArrow || upArrow) {
+      // The caret offset gives a location in the upper left hand corner of
+      // the caret so the middle of the line above is a half line above that
+      // point and the line below is 1.5 lines below that point.
+      final double plh = _textPainter.preferredLineHeight;
+      final double verticalOffset = upArrow ? -0.5 * plh : 1.5 * plh;
+
+      final Offset caretOffset = _textPainter.getOffsetForCaret(new TextPosition(offset: _extentOffset), _caretPrototype);
+      final Offset caretOffsetTranslated = caretOffset.translate(0.0, verticalOffset);
+      final TextPosition position = _textPainter.getPositionForOffset(caretOffsetTranslated);
+
+      // To account for the possibility where the user vertically highlights
+      // all the way to the top or bottom of the text, we hold the previous
+      // cursor location. This allows us to restore to this position in the
+      // case that the user wants to unhighlight some text.
+      if (position.offset == _extentOffset){
+        if (downArrow)
+          newOffset = text.text.length;
+        else if (upArrow)
+          newOffset = 0;
+        _resetCursor = shift;
+      } else if (_resetCursor && shift) {
+        newOffset = _previousCursorLocation;
+        _resetCursor = false;
+      } else {
+        newOffset = position.offset;
+        _previousCursorLocation = newOffset;
+      }
+    }
+    return newOffset;
+  }
+
+  // Simply handles the horizontal arrows
+  int _handleHorizontalArrows(bool rightArrow, bool leftArrow, bool shift, int newOffset) {
+    // Set the new offset to be +/- 1 depending on which arrow is pressed
+    // If shift is down, we also want to update the previous cursor location
+    if (rightArrow && _extentOffset < text.text.length) {
+      newOffset += 1;
+      if (shift)
+        _previousCursorLocation += 1;
+    }
+    if (leftArrow && _extentOffset > 0) {
+      newOffset -= 1;
+      if (shift)
+        _previousCursorLocation -= 1;
+    }
+    return newOffset;
+  }
+
+  // Handles full word traversal using control.
+  int _handleControl(bool rightArrow, bool leftArrow, bool ctrl, int newOffset) {
+    // If control is pressed, we will decide which way to look for a word
+    // based on which arrow is pressed.
+    if (ctrl) {
```

Comment:
since this method only does something if ctrl is true, it might be clearer (and would certainly be more efficient) to only call it when ctrl is true, and not have it deal with ctrl at all.